tardis/energy_input/util.py

```diff
@@ -0,0 +1,333 @@
+    float
+        Klein-Nishina solution
+    """
+    kappa = kappa_calculation(energy)
+    return (
+        R_ELECTRON.value
+        / 2
+        * (1.0 + kappa * (1.0 - np.cos(theta_C))) ** -2.0
+        * (
+            1.0
+            + np.cos(theta_C) ** 2.0
+            + (kappa ** 2.0 * (1.0 - np.cos(theta_C)) ** 2.0)
+            / (1.0 + kappa * (1.0 - np.cos(theta_C)))
+        )
+    )
+
+
+def compton_theta_distribution(energy, sample_resolution=100):
+    """
+    Calculates the cumulative distribution function of theta angles
+    for Compton Scattering
+
+    Parameters
+    ----------
+    energy : float
+    sample_resolution : int
+
+    Returns
+    -------
+    theta_angles : One dimensional Numpy array, dtype float
+    norm_theta_distribution : One dimensional Numpy array, dtype float
+
+    """
+    theta_angles = np.linspace(0, np.pi, sample_resolution)
+
+    theta_distribution = np.cumsum(klein_nishina(energy, theta_angles))
+    norm_theta_distribution = theta_distribution / np.max(theta_distribution)
+
+    return theta_angles, norm_theta_distribution
+
+
+def get_random_theta_gamma_ray():
+    """Get a random theta direction between 0 and pi
+    Returns
+    -------
+    float
+        Random theta direction
+    """
+    return np.arccos(1.0 - 2.0 * np.random.random())
+
+
+def get_random_phi_gamma_ray():
+    """Get a random phi direction between 0 and 2 * pi
+
+    Returns
+    -------
+    float
+        Random phi direction
+    """
+    return 2.0 * np.pi * np.random.random()
+
+
+def convert_half_life_to_astropy_units(half_life_string):
+    """Converts input half-life to use astropy units
+
+    Parameters
+    ----------
+    half_life_string : string
+        Half-life as a string
+
+    Returns
+    -------
+    astropy unit
+        Half-life in seconds
+    """
+    [value, unit] = half_life_string.split(" ")
+    try:
+        [nominal_value, magnitude] = value.split("×")
+        [base, exponent] = magnitude.split("+")
+        nominal_value = float(nominal_value) * float(base) ** float(exponent)
+    except ValueError:
+        nominal_value = float(value)
+    if unit == "y":
+        unit = "yr"
+    half_life_with_unit = nominal_value * u.Unit(unit)
+    return half_life_with_unit.to(u.s)
+
+
+def calculate_energy_per_mass(energy, mass):
+    """Calculates energy in erg / g
+
+    Parameters
+    ----------
+    energy : float64
+        Energy in KeV
+    mass : float64
+        Mass in grams
+
+    Returns
+    -------
+    float64
+        Energy density in erg / g
+    """
+    return energy * KEV2ERG / mass
+
+
+def normalize(vector):
+    """
+    Normalizes a vector in cartesian coordinates
+
+    Parameters
+    ----------
+    vector : One-dimensional Numpy Array, dtype float
+
+    Returns
+    -------
+    normalized_vector : One-dimensional Numpy Array, dtype float
+    """
+    normalized_vector = vector / np.sqrt(
+        vector[0] ** 2 + vector[1] ** 2 + vector[2] ** 2
```

Comment:
this should be the array squared but probably better `normalized_v = v/np.linalg.norm(v)`